File: src/algorithms/trees/manipulation/distribute-coins/sources/distribute-coins_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.dirname(__file__))
module = importlib.import_module("distribute-coins")
BinaryNode = module.BinaryNode
distribute_coins = module.distribute_coins


def make_node(value, left=None, right=None):
    node = BinaryNode(value)
    node.left = left
    node.right = right
    return node


def test_null_root_returns_zero():
    assert distribute_coins(None) == 0


def test_single_node_one_coin_returns_zero():
    assert distribute_coins(make_node(1)) == 0


def test_two_node_root_has_two_coins():
    root = make_node(2, make_node(0))
    assert distribute_coins(root) == 1


def test_root_three_coins_two_children_zero():
    root = make_node(3, make_node(0), make_node(0))
    assert distribute_coins(root) == 2


def test_all_coins_at_deep_leaf():
    root = make_node(0, make_node(0, make_node(3), None), make_node(0))
    assert distribute_coins(root) == 4


if __name__ == "__main__":
    test_null_root_returns_zero()
    test_single_node_one_coin_returns_zero()
    test_two_node_root_has_two_coins()
    test_root_three_coins_two_children_zero()
    test_all_coins_at_deep_leaf()
    print("All tests passed!")

File: src/algorithms/trees/manipulation/distribute-coins/__tests__/distribute-coins_test.rs
include!("../sources/distribute-coins.rs");

#[cfg(test)]
mod tests {
    use super::*;

    fn make_node(value: i32, left: Option<Box<BinaryNode>>, right: Option<Box<BinaryNode>>) -> Option<Box<BinaryNode>> {
        Some(Box::new(BinaryNode { value, left, right }))
    }

    fn leaf(value: i32) -> Option<Box<BinaryNode>> {
        make_node(value, None, None)
    }

    #[test]
    fn test_null_root_returns_zero() {
        assert_eq!(distribute_coins(&None), 0);
    }

    #[test]
    fn test_single_node_one_coin() {
        assert_eq!(distribute_coins(&leaf(1)), 0);
    }

    #[test]
    fn test_two_node_root_has_two_coins() {
        let root = make_node(2, leaf(0), None);
        assert_eq!(distribute_coins(&root), 1);
    }

    #[test]
    fn test_root_three_coins_two_zero_children() {
        let root = make_node(3, leaf(0), leaf(0));
        assert_eq!(distribute_coins(&root), 2);
    }

    #[test]
    fn test_all_coins_at_deep_leaf() {
        let root = make_node(0, make_node(0, leaf(3), None), leaf(0));
        assert_eq!(distribute_coins(&root), 4);
    }
}
